### PR TITLE
add support for tensorflow

### PR DESCRIPTION
**Pull request by madhur-tandon**
1. `save` and `load` for `TFTensorWriter` and `TFTensorReader` respectively aren't implemented yet.
2. `save` and `load` for `TFKerasModelIO` aren't implemented yet.
3. When `complex_net` is used in tests, it fails, probably because of using a list of tensors (just like the varargs in PyTorch which we removed)
4. Tests with `simple_net` fail because of `save/load` stuff not being implemented which is expected.
5. Sometimes, `argspec.args` is `[]` i.e. an empty list -- hence the `skip_first` flag also checks for it otherwise it will be index out of bounds for `Signature.from_method(...)`

### CODE REVIEW

File: tests/contrib/test_tensorflow.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+
+@pytest.fixture
+def tftt_3d(tensor_data):
+    return DataAnalyzer.analyze(
+        tf.tile(tf.expand_dims(tensor_data, -1), [1, 1, 20])
+    )
+
+
+@pytest.mark.skipif(
```
Comment (mike0sv) [inline tests/contrib/test_tensorflow.py:37]: not needed since we got correct env in tests

File: tests/contrib/test_tensorflow.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    tf.__version__.split(".")[0] != "2", reason="requires tensorflow 2.x"
+)
+def test_feed_dict_type__self_serialization(tftt):
+    from mlem.contrib.tensorflow import TFTensorDataType
```
Comment (mike0sv) [inline tests/contrib/test_tensorflow.py:41]: why local import?

File: tests/contrib/test_tensorflow.py (hunk 1 above)
Comment (madhur-tandon) [reply tests/contrib/test_tensorflow.py]: thanks for removing

File: mlem/contrib/tensorflow.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        raise NotImplementedError
+
+
+class TFKerasModelIO(ModelIO):
+    """
+    :class:`.ModelIO` implementation for Tensorflow Keras models (:class:`tensorflow.keras.Model` objects)
+    """
+
+    type: ClassVar[str] = "tensorflow_io"
+    model_file_name = "model.tf"
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply mlem/contrib/tensorflow.py]: duplicate

File: tests/contrib/test_tensorflow.py (hunk 2 above)
Comment (madhur-tandon) [reply tests/contrib/test_tensorflow.py]: fixed

File: mlem/contrib/tensorflow.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                f"expected: {getattr(tf, self.dtype)}"
+            )
+        self._check_shape(instance, SerializationError)
+        return instance.numpy().tolist()
+
+    def deserialize(self, obj):
+        try:
+            ret = tf.convert_to_tensor(obj, dtype=getattr(tf, self.dtype))
+        except (ValueError, TypeError):
+            raise DeserializationError(  # pylint: disable=W0707
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply mlem/contrib/tensorflow.py]: done

File: mlem/contrib/tensorflow.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+    def _check_shape(self, tensor, exc_type):
+        if tuple(tensor.shape)[1:] != self.shape[1:]:
+            raise exc_type(
+                f"given tensor is of shape: {(None,) + tuple(tensor.shape)[1:]}, expected: {self.shape}"
+            )
+
+    def serialize(self, instance: tf.Tensor):
+        self.check_type(instance, tf.Tensor, SerializationError)
+        if instance.dtype is not getattr(tf, self.dtype):
```
Comment (mike0sv) [inline mlem/contrib/tensorflow.py:55]: add proprety for `getattr(tf, self.dtype)`?

File: mlem/contrib/tensorflow.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    def get_requirements(self) -> Requirements:
+        return Requirements.new([InstallableRequirement.from_module(tf)])
+
+    def get_writer(
+        self, project: str = None, filename: str = None, **kwargs
+    ) -> DataWriter:
+        return TFTensorWriter(**kwargs)
+
+    def get_model(self, prefix: str = ""):
+        raise NotImplementedError
```
Comment (mike0sv) [inline mlem/contrib/tensorflow.py:83]: you should implement this, take numpy for reference

File: mlem/contrib/tensorflow.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+    @classmethod
+    def process(cls, obj: tf.Tensor, **kwargs) -> DataType:
+        return TFTensorDataType(
+            shape=tuple(obj.shape),
```
Comment (mike0sv) [inline mlem/contrib/tensorflow.py:88]: is the first dimension batch size? in that case it should be dynamic, represented by `None`

File: mlem/contrib/tensorflow.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    ):
+        if self.save_format == "h5":
+            if len(artifacts) != 1:
+                raise ValueError(
+                    "Invalid artifacts: should have only one file"
+                )
+            with artifacts[self.art_name].open() as f:
+                return tf.keras.models.load_model(h5py.File(f))
+
+        if self.save_format == "tf":
```
Comment (mike0sv) [inline mlem/contrib/tensorflow.py:159]: add error for unknown save format

File: mlem/contrib/tensorflow.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    def save_model(self, model: tf.keras.Model, path: str):
+        if self.save_format is None:
+            self.save_format = "tf" if is_custom_net(model) else "h5"
+        model.save(path, save_format=self.save_format)
+
+    def load(  # pylint:disable=inconsistent-return-statements
+        self, artifacts: Artifacts
+    ):
+        if self.save_format == "h5":
+            if len(artifacts) != 1:
```
Comment (mike0sv) [inline mlem/contrib/tensorflow.py:152]: better to check `self.art_name in artifacts`

File: setup.py [hunk 10]
```diff
@@ -147,13 +151,15 @@
             "model_io.pickle = mlem.contrib.callable:PickleModelIO",
             "model_io.xgboost_io = mlem.contrib.xgboost:XGBoostModelIO",
             "model_io.torch_io = mlem.contrib.torch:TorchModelIO",
+            "model_io.tensorflow_io = mlem.contrib.tensorflow:TFKerasModelIO",
```
Comment (mike0sv) [inline setup.py:154]: tf_keras?

File: setup.py [hunk 11]
```diff
@@ -121,19 +122,22 @@
...
             "data_reader.pandas_series = mlem.contrib.pandas:PandasSeriesReader",
             "data_reader.torch = mlem.contrib.torch:TorchTensorReader",
+            "data_reader.tensorflow = mlem.contrib.tensorflow:TFTensorReader",
             "data_type.dataframe = mlem.contrib.pandas:DataFrameType",
             "data_type.lightgbm = mlem.contrib.lightgbm:LightGBMDataType",
             "data_type.ndarray = mlem.contrib.numpy:NumpyNdarrayType",
             "data_type.number = mlem.contrib.numpy:NumpyNumberType",
             "data_type.series = mlem.contrib.pandas:SeriesType",
             "data_type.torch = mlem.contrib.torch:TorchTensorDataType",
+            "data_type.tensorflow = mlem.contrib.tensorflow:TFTensorDataType",
```
Comment (mike0sv) [inline setup.py:132]: tf_tensor?

File: setup.py [hunk 12]
```diff
@@ -147,13 +151,15 @@
...
             "model_io.torch_io = mlem.contrib.torch:TorchModelIO",
+            "model_io.tensorflow_io = mlem.contrib.tensorflow:TFKerasModelIO",
             "model_type.callable = mlem.contrib.callable:CallableModelType",
             "model_type.catboost = mlem.contrib.catboost:CatBoostModel",
             "model_type.lightgbm = mlem.contrib.lightgbm:LightGBMModel",
             "model_type.sklearn = mlem.contrib.sklearn:SklearnModel",
             "model_type.sklearn_pipeline = mlem.contrib.sklearn:SklearnPipelineType",
             "model_type.xgboost = mlem.contrib.xgboost:XGBoostModel",
             "model_type.torch = mlem.contrib.torch:TorchModel",
+            "model_type.tensorflow = mlem.contrib.tensorflow:TFKerasModel",
```
Comment (mike0sv) [inline setup.py:162]: tf_keras?

Comment (madhur-tandon) [reply setup.py]: done

File: setup.py (hunk 11 above)
Comment (madhur-tandon) [reply setup.py]: done

File: setup.py (hunk 10 above)
Comment (madhur-tandon) [reply setup.py]: done

File: mlem/contrib/tensorflow.py (hunk 9 above)
Comment (madhur-tandon) [reply mlem/contrib/tensorflow.py]: done

File: mlem/contrib/tensorflow.py (hunk 8 above)
Comment (madhur-tandon) [reply mlem/contrib/tensorflow.py:182]: done

File: mlem/contrib/tensorflow.py (hunk 7 above)
Comment (madhur-tandon) [reply mlem/contrib/tensorflow.py]: done

File: mlem/contrib/tensorflow.py (hunk 6 above)
Comment (madhur-tandon) [reply mlem/contrib/tensorflow.py]: done

File: mlem/contrib/tensorflow.py (hunk 5 above)
Comment (madhur-tandon) [reply mlem/contrib/tensorflow.py]: done

File: .github/workflows/check-test-release.yml [hunk 13]
```diff
@@ -78,7 +78,7 @@ jobs:
       if: matrix.python != '3.7'
     - name: Run tests
       timeout-minutes: 40
-      run: pytest
+      run: pytest -m "not long" && pytest -m long
```
[the remark this answers is not in the document]
Comment (aguschin) [reply .github/workflows/check-test-release.yml:81]:
```suggestion
      run: pytest
```

File: mlem/ext.py [hunk 14]
```diff
@@ -92,8 +92,7 @@ class ExtensionLoader:
         Extension("mlem.contrib.numpy", ["numpy"], False),
         Extension("mlem.contrib.pandas", ["pandas"], False),
         Extension("mlem.contrib.sklearn", ["sklearn"], False),
-        # Extension('mlem.contrib.tensorflow', ['tensorflow'], False, is_tf_v1),
-        # Extension('mlem.contrib.tensorflow_v2', ['tensorflow'], False, is_tf_v2),
+        Extension("mlem.contrib.tensorflow", ["tensorflow"], False),
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply mlem/ext.py:95]: only `v2`

**Review by aguschin — APPROVED** [review]
LGTM! Great work @madhur-tandon !

**Pull request merged**